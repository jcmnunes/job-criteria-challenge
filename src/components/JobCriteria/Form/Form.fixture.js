export default {
  formState: {
    experience: '',
    educations: [],
    hours: {
      min: '',
      max: '',
    },
  },
};
