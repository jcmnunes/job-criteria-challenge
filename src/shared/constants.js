export const EDUCATIONS = [
  {
    name: 'bachelor',
    label: 'Bachelor / Graduate',
  },
  {
    name: 'gcse',
    label: 'GCSE / A-Level / Highschool / GED',
  },
  {
    name: 'master',
    label: 'Master / Post-Graduate / PhD',
  },
  {
    name: 'vocational',
    label: 'Vocational / Diploma / Associates degree',
  },
];

export const LABELS = {
  experience: 'A minimum No. years of experience',
  educations: 'Level of education',
  hours: 'No. of working hours (per week)',
};

export const ERRORS = {
  experienceError: '',
  educationsError: '',
  hoursMinError: '',
  hoursMaxError: '',
};
