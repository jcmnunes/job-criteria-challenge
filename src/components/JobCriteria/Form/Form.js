import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InputField from '../../UI/InputField';
import FormHeader from '../../UI/FormHeader';
import CheckboxesField from '../../UI/CheckboxesField';
import RangeInputField from '../../UI/RangeInputField';
import FormFooter from '../../UI/FormFooter';
import Button from '../../UI/Button';
import theme from '../../../styles/theme';
import { EDUCATIONS, LABELS, ERRORS } from '../../../shared/constants';

class Form extends Component {
  state = {
    ...ERRORS,
    isSubmitting: false,
  };

  handleSubmit = e => {
    e.preventDefault();
    this.validate(() => {
      if (this.isFormValid()) {
        this.setState({ isSubmitting: true });
        // A timeout here just for fun
        setTimeout(() => {
          this.setState({ isSubmitting: false });
          this.props.handleSubmitSuccess();
        }, 500);
      }
    });
  };

  validate = cb => {
    const { formState } = this.props;
    const { experience, educations, hours } = formState;
    const newErrors = { ...ERRORS };
    if (experience.length === 0) {
      newErrors.experienceError = 'The field is required.';
    }
    if (educations.length === 0) {
      newErrors.educationsError = 'The field is required.';
    }
    if (hours.min.length === 0) {
      newErrors.hoursMinError = 'The field is required.';
    }
    if (hours.max.length === 0) {
      newErrors.hoursMaxError = 'The field is required.';
    }
    if (
      hours.max.length !== 0 &&
      hours.min.length !== 0 &&
      !(parseInt(hours.max, 10) > parseInt(hours.min, 10))
    ) {
      newErrors.hoursMinError = `Minimum value must be less than ${formState.hours.max}`;
    }
    return this.setState({ ...newErrors }, cb);
  };

  isFormValid = () => {
    const { formState } = this.props;
    const { experience, educations, hours } = formState;
    const { experienceError, educationsError, hoursMinError, hoursMaxError } = this.state;
    return (
      experienceError.length === 0 &&
      educationsError.length === 0 &&
      hoursMinError.length === 0 &&
      hoursMaxError.length === 0 &&
      String(experience).length > 0 &&
      educations.length > 0 &&
      hours.min.length > 0 &&
      hours.max.length > 0
    );
  };

  handleOnChange = e => {
    const { experienceError, educationsError, hoursMinError, hoursMaxError } = this.state;
    if (
      experienceError.length > 0 ||
      educationsError.length > 0 ||
      hoursMinError.length > 0 ||
      hoursMaxError.length > 0
    ) {
      this.setState({ ...ERRORS });
    }

    this.props.onChange(e);
  };

  render() {
    const { formState } = this.props;
    const {
      experienceError,
      educationsError,
      hoursMinError,
      hoursMaxError,
      isSubmitting,
    } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <FormHeader
          title="Job criteria"
          helper="We will use the data we collect here to better target your desired audience"
          badge={{ value: '1', color: theme.purple500 }}
        />
        <InputField
          name="experience"
          type="number"
          label={LABELS.experience}
          value={formState.experience}
          onChange={this.handleOnChange}
          min="0"
          error={experienceError}
        />
        <CheckboxesField
          label={LABELS.educations}
          checkboxes={EDUCATIONS}
          state={formState.educations}
          onChange={this.handleOnChange}
          error={educationsError}
        />
        <RangeInputField
          label={LABELS.hours}
          state={formState.hours}
          onChange={this.handleOnChange}
          minName="min"
          maxName="max"
          minError={hoursMinError}
          maxError={hoursMaxError}
          error={hoursMinError || hoursMaxError}
          min="1"
        />
        <FormFooter>
          <Button type="submit" loading={isSubmitting}>
            Save
          </Button>
        </FormFooter>
      </form>
    );
  }
}

Form.propTypes = {
  formState: PropTypes.shape({
    experience: PropTypes.string.isRequired,
    educations: PropTypes.array.isRequired,
    hours: PropTypes.shape({
      min: PropTypes.string.isRequired,
      max: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  handleSubmitSuccess: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Form;
