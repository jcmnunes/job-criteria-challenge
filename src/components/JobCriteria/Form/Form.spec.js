import React from 'react';
import { shallow } from 'enzyme';

import Form from './Form';
import fixture from './Form.fixture';
import { ERRORS } from '../../../shared/constants';

describe('Components', () => {
  const props = {
    ...fixture,
    handleSubmitSuccess: jest.fn(),
    onChange: jest.fn(),
  };

  const formValidProps = {
    ...props,
    formState: {
      ...props.formState,
      experience: '5',
      educations: ['bachelor'],
      hours: {
        min: '1',
        max: '2',
      },
    },
  };

  jest.useFakeTimers();

  describe('Form', () => {
    it('should render as expected, without crashing', () => {
      const wrapper = shallow(<Form {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should not submit the form if it is not valid', () => {
      const wrapper = shallow(<Form {...props} />);

      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
      });
      expect(props.handleSubmitSuccess).not.toHaveBeenCalled();
    });

    it('should submit the form if the form is valid', () => {
      const newProps = {
        ...props,
        formState: {
          ...props.formState,
          experience: '5',
          educations: ['bachelor'],
          hours: {
            min: '1',
            max: '2',
          },
        },
      };
      const wrapper = shallow(<Form {...newProps} />);

      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
      });
      jest.runAllTimers();

      expect(props.handleSubmitSuccess).toHaveBeenCalled();
    });

    it('should set the error message', () => {
      const wrapper = shallow(<Form {...props} />);
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
      });

      expect(wrapper.state().experienceError).toBe('The field is required.');
      expect(wrapper.state().educationsError).toBe('The field is required.');
      expect(wrapper.state().hoursMinError).toBe('The field is required.');
    });

    it('should set the error message (hours errors)', () => {
      const newProps = {
        ...props,
        formState: {
          ...props.formState,
          experience: '',
          educations: [],
          hours: {
            min: '10',
            max: '5',
          },
        },
      };
      const wrapper = shallow(<Form {...newProps} />);
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
      });

      expect(wrapper.state().hoursMinError).toBe(`Minimum value must be less than ${5}`);
    });

    it('should check that form is invalid (isValid method)', () => {
      const wrapper = shallow(<Form {...props} />);
      const result = wrapper.instance().isFormValid();

      expect(result).toBeFalsy();
    });

    it('should check that form is valid (isValid method)', () => {
      const wrapper = shallow(<Form {...formValidProps} />);
      const result = wrapper.instance().isFormValid();

      expect(result).toBeTruthy();
    });

    it('should reset errors on input change', () => {
      const wrapper = shallow(<Form {...props} />);
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {},
      });

      expect(wrapper.state()).not.toEqual({ ...ERRORS, isSubmitting: false });
      wrapper.instance().handleOnChange();
      expect(wrapper.state()).toEqual({ ...ERRORS, isSubmitting: false });
    });
  });
});
