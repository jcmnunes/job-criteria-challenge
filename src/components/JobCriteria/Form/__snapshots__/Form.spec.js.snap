// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Form should render as expected, without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    formState={
      Object {
        "educations": Array [],
        "experience": "",
        "hours": Object {
          "max": "",
          "min": "",
        },
      }
    }
    handleSubmitSuccess={[MockFunction]}
    onChange={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <FormHeader
          badge={
            Object {
              "color": "#A986E6",
              "value": "1",
            }
          }
          helper="We will use the data we collect here to better target your desired audience"
          title="Job criteria"
        />,
        <InputField
          error=""
          label="A minimum No. years of experience"
          min="0"
          name="experience"
          onChange={[Function]}
          type="number"
          value=""
        />,
        <CheckboxesField
          checkboxes={
            Array [
              Object {
                "label": "Bachelor / Graduate",
                "name": "bachelor",
              },
              Object {
                "label": "GCSE / A-Level / Highschool / GED",
                "name": "gcse",
              },
              Object {
                "label": "Master / Post-Graduate / PhD",
                "name": "master",
              },
              Object {
                "label": "Vocational / Diploma / Associates degree",
                "name": "vocational",
              },
            ]
          }
          error=""
          label="Level of education"
          onChange={[Function]}
          state={Array []}
        />,
        <RangeInputField
          error=""
          label="No. of working hours (per week)"
          leftError=""
          leftName="min"
          onChange={[Function]}
          rightError=""
          rightName="max"
          state={
            Object {
              "max": "",
              "min": "",
            }
          }
        />,
        <ForwardRef>
          <Button
            disabled={false}
            loading={false}
            type="submit"
          >
            Save
          </Button>
        </ForwardRef>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "badge": Object {
            "color": "#A986E6",
            "value": "1",
          },
          "helper": "We will use the data we collect here to better target your desired audience",
          "title": "Job criteria",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "error": "",
          "label": "A minimum No. years of experience",
          "min": "0",
          "name": "experience",
          "onChange": [Function],
          "type": "number",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "checkboxes": Array [
            Object {
              "label": "Bachelor / Graduate",
              "name": "bachelor",
            },
            Object {
              "label": "GCSE / A-Level / Highschool / GED",
              "name": "gcse",
            },
            Object {
              "label": "Master / Post-Graduate / PhD",
              "name": "master",
            },
            Object {
              "label": "Vocational / Diploma / Associates degree",
              "name": "vocational",
            },
          ],
          "error": "",
          "label": "Level of education",
          "onChange": [Function],
          "state": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "error": "",
          "label": "No. of working hours (per week)",
          "leftError": "",
          "leftName": "min",
          "onChange": [Function],
          "rightError": "",
          "rightName": "max",
          "state": Object {
            "max": "",
            "min": "",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Button
            disabled={false}
            loading={false}
            type="submit"
          >
            Save
          </Button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Save",
            "disabled": false,
            "loading": false,
            "type": "submit",
          },
          "ref": null,
          "rendered": "Save",
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-gZMcBi",
            "isStatic": true,
            "rules": Array [
              "
  display: flex;
  justify-content: flex-end;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-gZMcBi",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <FormHeader
            badge={
              Object {
                "color": "#A986E6",
                "value": "1",
              }
            }
            helper="We will use the data we collect here to better target your desired audience"
            title="Job criteria"
          />,
          <InputField
            error=""
            label="A minimum No. years of experience"
            min="0"
            name="experience"
            onChange={[Function]}
            type="number"
            value=""
          />,
          <CheckboxesField
            checkboxes={
              Array [
                Object {
                  "label": "Bachelor / Graduate",
                  "name": "bachelor",
                },
                Object {
                  "label": "GCSE / A-Level / Highschool / GED",
                  "name": "gcse",
                },
                Object {
                  "label": "Master / Post-Graduate / PhD",
                  "name": "master",
                },
                Object {
                  "label": "Vocational / Diploma / Associates degree",
                  "name": "vocational",
                },
              ]
            }
            error=""
            label="Level of education"
            onChange={[Function]}
            state={Array []}
          />,
          <RangeInputField
            error=""
            label="No. of working hours (per week)"
            leftError=""
            leftName="min"
            onChange={[Function]}
            rightError=""
            rightName="max"
            state={
              Object {
                "max": "",
                "min": "",
              }
            }
          />,
          <ForwardRef>
            <Button
              disabled={false}
              loading={false}
              type="submit"
            >
              Save
            </Button>
          </ForwardRef>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "badge": Object {
              "color": "#A986E6",
              "value": "1",
            },
            "helper": "We will use the data we collect here to better target your desired audience",
            "title": "Job criteria",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "error": "",
            "label": "A minimum No. years of experience",
            "min": "0",
            "name": "experience",
            "onChange": [Function],
            "type": "number",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "checkboxes": Array [
              Object {
                "label": "Bachelor / Graduate",
                "name": "bachelor",
              },
              Object {
                "label": "GCSE / A-Level / Highschool / GED",
                "name": "gcse",
              },
              Object {
                "label": "Master / Post-Graduate / PhD",
                "name": "master",
              },
              Object {
                "label": "Vocational / Diploma / Associates degree",
                "name": "vocational",
              },
            ],
            "error": "",
            "label": "Level of education",
            "onChange": [Function],
            "state": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "error": "",
            "label": "No. of working hours (per week)",
            "leftError": "",
            "leftName": "min",
            "onChange": [Function],
            "rightError": "",
            "rightName": "max",
            "state": Object {
              "max": "",
              "min": "",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Button
              disabled={false}
              loading={false}
              type="submit"
            >
              Save
            </Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Save",
              "disabled": false,
              "loading": false,
              "type": "submit",
            },
            "ref": null,
            "rendered": "Save",
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-gZMcBi",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  justify-content: flex-end;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-gZMcBi",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
